// expandjs/lib/languages.js
/**
 * @license
 * Copyright (c) 2017 The expand.js authors. All rights reserved.
 * This code may only be used under the BSD style license found at https://expandjs.github.io/LICENSE.txt
 * The complete set of authors may be found at https://expandjs.github.io/AUTHORS.txt
 * The complete set of contributors may be found at https://expandjs.github.io/CONTRIBUTORS.txt
 */

// Exporting
module.exports = {
    af_NA: 'af',
    af_ZA: 'af',
    ak_GH: 'ak',
    am_ET: 'am',
    ar_AE: 'ar',
    ar_BH: 'ar',
    ar_DJ: 'ar',
    ar_DZ: 'ar',
    ar_EG: 'ar',
    ar_EH: 'ar',
    ar_ER: 'ar',
    ar_IL: 'ar',
    ar_IQ: 'ar',
    ar_JO: 'ar',
    ar_KM: 'ar',
    ar_KW: 'ar',
    ar_LB: 'ar',
    ar_LY: 'ar',
    ar_MA: 'ar',
    ar_MR: 'ar',
    ar_OM: 'ar',
    ar_PS: 'ar',
    ar_QA: 'ar',
    ar_SA: 'ar',
    ar_SD: 'ar',
    ar_SO: 'ar',
    ar_SS: 'ar',
    ar_SY: 'ar',
    ar_TD: 'ar',
    ar_TN: 'ar',
    ar_YE: 'ar',
    as_IN: 'as',
    be_BY: 'be',
    bg_BG: 'bg',
    bn_BD: 'bn',
    bn_IN: 'bn',
    bo_CN: 'bo',
    bo_IN: 'bo',
    br_FR: 'br',
    bs_BA: 'bs',
    ca_AD: 'ca',
    ca_ES: 'ca',
    ca_FR: 'ca',
    ca_IT: 'ca',
    cs_CZ: 'cs',
    cy_GB: 'cy',
    da_DK: 'da',
    da_GL: 'da',
    de_AT: 'de',
    de_BE: 'de',
    de_CH: 'de',
    de_DE: 'de',
    de_LI: 'de',
    de_LU: 'de',
    dz_BT: 'dz',
    ee_GH: 'ee',
    ee_TG: 'ee',
    el_CY: 'el',
    el_GR: 'el',
    en_AD: 'en',
    en_AG: 'en',
    en_AI: 'en',
    en_AL: 'en',
    en_AS: 'en',
    en_AT: 'en',
    en_AU: 'en',
    en_BA: 'en',
    en_BB: 'en',
    en_BE: 'en',
    en_BM: 'en',
    en_BS: 'en',
    en_BW: 'en',
    en_BZ: 'en',
    en_CA: 'en',
    en_CC: 'en',
    en_CH: 'en',
    en_CK: 'en',
    en_CM: 'en',
    en_CX: 'en',
    en_CY: 'en',
    en_CZ: 'en',
    en_DE: 'en',
    en_DG: 'en',
    en_DK: 'en',
    en_DM: 'en',
    en_EE: 'en',
    en_ER: 'en',
    en_ES: 'en',
    en_FI: 'en',
    en_FJ: 'en',
    en_FK: 'en',
    en_FM: 'en',
    en_FR: 'en',
    en_GB: 'en',
    en_GD: 'en',
    en_GG: 'en',
    en_GH: 'en',
    en_GI: 'en',
    en_GM: 'en',
    en_GR: 'en',
    en_GU: 'en',
    en_GY: 'en',
    en_HK: 'en',
    en_HR: 'en',
    en_HU: 'en',
    en_IE: 'en',
    en_IL: 'en',
    en_IM: 'en',
    en_IN: 'en',
    en_IO: 'en',
    en_IS: 'en',
    en_IT: 'en',
    en_JE: 'en',
    en_JM: 'en',
    en_KE: 'en',
    en_KI: 'en',
    en_KN: 'en',
    en_KY: 'en',
    en_LC: 'en',
    en_LR: 'en',
    en_LS: 'en',
    en_LT: 'en',
    en_LU: 'en',
    en_LV: 'en',
    en_ME: 'en',
    en_MG: 'en',
    en_MH: 'en',
    en_MO: 'en',
    en_MP: 'en',
    en_MS: 'en',
    en_MT: 'en',
    en_MU: 'en',
    en_MW: 'en',
    en_MY: 'en',
    en_NA: 'en',
    en_NF: 'en',
    en_NG: 'en',
    en_NL: 'en',
    en_NO: 'en',
    en_NR: 'en',
    en_NU: 'en',
    en_NZ: 'en',
    en_PG: 'en',
    en_PH: 'en',
    en_PK: 'en',
    en_PL: 'en',
    en_PN: 'en',
    en_PR: 'en',
    en_PT: 'en',
    en_PW: 'en',
    en_RO: 'en',
    en_RU: 'en',
    en_RW: 'en',
    en_SB: 'en',
    en_SC: 'en',
    en_SD: 'en',
    en_SE: 'en',
    en_SG: 'en',
    en_SH: 'en',
    en_SI: 'en',
    en_SK: 'en',
    en_SL: 'en',
    en_SS: 'en',
    en_SX: 'en',
    en_SZ: 'en',
    en_TC: 'en',
    en_TK: 'en',
    en_TO: 'en',
    en_TR: 'en',
    en_TT: 'en',
    en_TV: 'en',
    en_TZ: 'en',
    en_UG: 'en',
    en_UM: 'en',
    en_US: 'en',
    en_VC: 'en',
    en_VG: 'en',
    en_VI: 'en',
    en_VU: 'en',
    en_WS: 'en',
    en_ZA: 'en',
    en_ZM: 'en',
    en_ZW: 'en',
    es_AR: 'es',
    es_BO: 'es',
    es_CL: 'es',
    es_CO: 'es',
    es_CR: 'es',
    es_CU: 'es',
    es_DO: 'es',
    es_EA: 'es',
    es_EC: 'es',
    es_ES: 'es',
    es_GQ: 'es',
    es_GT: 'es',
    es_HN: 'es',
    es_IC: 'es',
    es_MX: 'es',
    es_NI: 'es',
    es_PA: 'es',
    es_PE: 'es',
    es_PH: 'es',
    es_PR: 'es',
    es_PY: 'es',
    es_SV: 'es',
    es_US: 'es',
    es_UY: 'es',
    es_VE: 'es',
    et_EE: 'et',
    eu_ES: 'eu',
    fa_AF: 'fa',
    fa_IR: 'fa',
    ff_CM: 'ff',
    ff_GN: 'ff',
    ff_MR: 'ff',
    ff_SN: 'ff',
    fi_FI: 'fi',
    fo_FO: 'fo',
    fr_BE: 'fr',
    fr_BF: 'fr',
    fr_BI: 'fr',
    fr_BJ: 'fr',
    fr_BL: 'fr',
    fr_CA: 'fr',
    fr_CD: 'fr',
    fr_CF: 'fr',
    fr_CG: 'fr',
    fr_CH: 'fr',
    fr_CI: 'fr',
    fr_CM: 'fr',
    fr_DJ: 'fr',
    fr_DZ: 'fr',
    fr_FR: 'fr',
    fr_GA: 'fr',
    fr_GF: 'fr',
    fr_GN: 'fr',
    fr_GP: 'fr',
    fr_GQ: 'fr',
    fr_HT: 'fr',
    fr_KM: 'fr',
    fr_LU: 'fr',
    fr_MA: 'fr',
    fr_MC: 'fr',
    fr_MF: 'fr',
    fr_MG: 'fr',
    fr_ML: 'fr',
    fr_MQ: 'fr',
    fr_MR: 'fr',
    fr_MU: 'fr',
    fr_NC: 'fr',
    fr_NE: 'fr',
    fr_PF: 'fr',
    fr_PM: 'fr',
    fr_RE: 'fr',
    fr_RW: 'fr',
    fr_SC: 'fr',
    fr_SN: 'fr',
    fr_SY: 'fr',
    fr_TD: 'fr',
    fr_TG: 'fr',
    fr_TN: 'fr',
    fr_VU: 'fr',
    fr_WF: 'fr',
    fr_YT: 'fr',
    fy_NL: 'fy',
    ga_IE: 'ga',
    gd_GB: 'gd',
    gl_ES: 'gl',
    gu_IN: 'gu',
    gv_IM: 'gv',
    he_IL: 'he',
    hi_IN: 'hi',
    hr_BA: 'hr',
    hr_HR: 'hr',
    hu_HU: 'hu',
    hy_AM: 'hy',
    id_ID: 'id',
    ig_NG: 'ig',
    ii_CN: 'ii',
    is_IS: 'is',
    it_CH: 'it',
    it_IT: 'it',
    it_SM: 'it',
    ja_JP: 'ja',
    jv_ID: 'jv',
    ka_GE: 'ka',
    ki_KE: 'ki',
    kl_GL: 'kl',
    km_KH: 'km',
    kn_IN: 'kn',
    ko_KP: 'ko',
    ko_KR: 'ko',
    kw_GB: 'kw',
    ky_KY: 'ky',
    lb_LU: 'lb',
    lg_UG: 'lg',
    ln_AO: 'ln',
    ln_CD: 'ln',
    ln_CF: 'ln',
    ln_CG: 'ln',
    lo_LA: 'lo',
    lt_LT: 'lt',
    lu_CD: 'lu',
    lv_LV: 'lv',
    me_ME: 'me',
    mg_MG: 'mg',
    mk_MK: 'mk',
    ml_IN: 'ml',
    mr_IN: 'mr',
    ms_MY: 'ms',
    mt_MT: 'mt',
    my_MM: 'my',
    nb_NO: 'nb',
    nb_SJ: 'nb',
    nd_ZW: 'nd',
    ne_IN: 'ne',
    ne_NP: 'ne',
    nl_AW: 'nl',
    nl_BE: 'nl',
    nl_BQ: 'nl',
    nl_CW: 'nl',
    nl_NL: 'nl',
    nl_SR: 'nl',
    nl_SX: 'nl',
    nn_NO: 'nn',
    om_ET: 'om',
    om_KE: 'om',
    or_IN: 'or',
    os_GE: 'os',
    os_RU: 'os',
    pl_PL: 'pl',
    ps_AF: 'ps',
    pt_AO: 'pt',
    pt_BR: 'pt',
    pt_CV: 'pt',
    pt_GW: 'pt',
    pt_MO: 'pt',
    pt_MZ: 'pt',
    pt_PT: 'pt',
    pt_ST: 'pt',
    pt_TL: 'pt',
    qu_BO: 'qu',
    qu_EC: 'qu',
    qu_PE: 'qu',
    rm_CH: 'rm',
    rn_BI: 'rn',
    ro_MD: 'ro',
    ro_RO: 'ro',
    ru_BY: 'ru',
    ru_KG: 'ru',
    ru_KZ: 'ru',
    ru_MD: 'ru',
    ru_RU: 'ru',
    ru_UA: 'ru',
    rw_RW: 'rw',
    se_FI: 'se',
    se_NO: 'se',
    se_SE: 'se',
    sg_CF: 'sg',
    si_LK: 'si',
    sk_SK: 'sk',
    sl_SI: 'sl',
    sn_ZW: 'sn',
    so_DJ: 'so',
    so_ET: 'so',
    so_KE: 'so',
    so_SO: 'so',
    sq_AL: 'sq',
    sq_MK: 'sq',
    sq_XK: 'sq',
    sr_RS: 'sr',
    su_ID: 'su',
    sv_AX: 'sv',
    sv_FI: 'sv',
    sv_SE: 'sv',
    sw_CD: 'sw',
    sw_KE: 'sw',
    sw_TZ: 'sw',
    sw_UG: 'sw',
    ta_IN: 'ta',
    ta_LK: 'ta',
    ta_MY: 'ta',
    ta_SG: 'ta',
    te_IN: 'te',
    th_TH: 'th',
    ti_ER: 'ti',
    ti_ET: 'ti',
    to_TO: 'to',
    tr_CY: 'tr',
    tr_TR: 'tr',
    tt_RU: 'tt',
    ug_CN: 'ug',
    uk_UA: 'uk',
    ur_IN: 'ur',
    ur_PK: 'ur',
    vi_VN: 'vi',
    yo_BJ: 'yo',
    yo_NG: 'yo',
    zu_ZA: 'zu'
};
