// expandjs/lib/locales.js
/**
 * @license
 * Copyright (c) 2017 The expand.js authors. All rights reserved.
 * This code may only be used under the BSD style license found at https://expandjs.github.io/LICENSE.txt
 * The complete set of authors may be found at https://expandjs.github.io/AUTHORS.txt
 * The complete set of contributors may be found at https://expandjs.github.io/CONTRIBUTORS.txt
 */

// Exporting
module.exports = {
    af: 'af_ZA',
    ak: 'ak_GH',
    am: 'am_ET',
    ar: 'ar_SA',
    as: 'as_IN',
    be: 'be_BY',
    bg: 'bg_BG',
    bn: 'bn_BD',
    bo: 'bo_CN',
    br: 'br_FR',
    bs: 'bs_BA',
    ca: 'ca_ES',
    cs: 'cs_CZ',
    cy: 'cy_GB',
    da: 'da_DK',
    de: 'de_DE',
    dz: 'dz_BT',
    ee: 'ee_GH',
    el: 'el_GR',
    en: 'en_US',
    es: 'es_ES',
    et: 'et_EE',
    eu: 'eu_ES',
    fa: 'fa_IR',
    ff: 'ff_CM',
    fi: 'fi_FI',
    fo: 'fo_FO',
    fr: 'fr_FR',
    fy: 'fy_NL',
    ga: 'ga_IE',
    gd: 'gd_GB',
    gl: 'gl_ES',
    gu: 'gu_IN',
    gv: 'gv_IM',
    he: 'he_IL',
    hi: 'hi_IN',
    hr: 'hr_HR',
    hu: 'hu_HU',
    hy: 'hy_AM',
    id: 'id_ID',
    ig: 'ig_NG',
    ii: 'ii_CN',
    is: 'is_IS',
    it: 'it_IT',
    ja: 'ja_JP',
    jv: 'jv_ID',
    ka: 'ka_GE',
    ki: 'ki_KE',
    kl: 'kl_GL',
    km: 'km_KH',
    kn: 'kn_IN',
    ko: 'ko_KR',
    kw: 'kw_GB',
    ky: 'ky_KY',
    lb: 'lb_LU',
    lg: 'lg_UG',
    ln: 'ln_CF',
    lo: 'lo_LA',
    lt: 'lt_LT',
    lu: 'lu_CD',
    lv: 'lv_LV',
    me: 'me_ME',
    mg: 'mg_MG',
    mk: 'mk_MK',
    ml: 'ml_IN',
    mr: 'mr_IN',
    ms: 'ms_MY',
    mt: 'mt_MT',
    my: 'my_MM',
    nb: 'nb_NO',
    nd: 'nd_ZW',
    ne: 'ne_NP',
    nl: 'nl_NL',
    nn: 'nn_NO',
    om: 'om_ET',
    or: 'or_IN',
    os: 'os_GE',
    pa: 'pa_IN',
    pl: 'pl_PL',
    ps: 'ps_AF',
    pt: 'pt_PT',
    qu: 'qu_EC',
    rm: 'rm_CH',
    rn: 'rn_BI',
    ro: 'ro_RO',
    ru: 'ru_RU',
    rw: 'rw_RW',
    sd: 'sd_PK',
    se: 'se_SE',
    sg: 'sg_CF',
    si: 'si_LK',
    sk: 'sk_SK',
    sl: 'sl_SI',
    sn: 'sn_ZW',
    so: 'so_SO',
    sq: 'sq_AL',
    sr: 'sr_RS',
    su: 'su_ID',
    sv: 'sv_SE',
    sw: 'sw_TZ',
    ta: 'ta_IN',
    te: 'te_IN',
    th: 'th_TH',
    ti: 'ti_ET',
    to: 'to_TO',
    tr: 'tr_TR',
    tt: 'tt_RU',
    ug: 'ug_CN',
    uk: 'uk_UA',
    ur: 'ur_PK',
    uz: 'uz_UZ',
    vi: 'vi_VN',
    yo: 'yo_NG',
    zu: 'zu_ZA'
};
